src/async_fuse/memfs/cache/block.rs

```diff
@@ -0,0 +1,306 @@
+//! Utilities for blocks.
+
+use std::{fmt::Formatter, sync::Arc};
+
+use aligned_utils::bytes::AlignedBytes;
+use clippy_utilities::OverflowArithmetic;
+use nix::sys::uio::IoVec;
+
+use crate::async_fuse::fuse::fuse_reply::{AsIoVec, CouldBeAsIoVecList};
+use crate::async_fuse::fuse::protocol::INum;
+
+/// Page Size
+const PAGE_SIZE: usize = 4096;
+
+/// A common coordinate to locate a block.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub struct BlockCoordinate(pub INum, pub usize);
+
+/// Make a partial slice of `u8` slice for debugging.
+fn make_partial_u8_slice(src: &[u8], start: usize, len: usize) -> String {
+    let end = start.overflow_add(len).min(src.len());
+    let show_length = end.overflow_sub(start);
+    let slice = src.get(start..end).unwrap_or(&[]);
+
+    let mut result = String::from("[");
+
+    for (i, &ch) in slice.iter().enumerate() {
+        result.push_str(format!("{ch}").as_str());
+        if i != show_length.overflow_sub(1) {
+            result.push_str(", ");
+        }
+    }
+
+    if end != src.len() {
+        result.push_str(", ...");
+    }
+
+    result.push(']');
+
+    result
+}
+
+/// The minimum unit of data in the storage layers.
+#[derive(Clone)]
+pub struct Block {
+    /// The underlying data of a block. Shared with `Arc`.
+    inner: Arc<AlignedBytes>,
+    /// A flag that if this block is dirty.
+    dirty: bool,
+}
+
+impl std::fmt::Debug for Block {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let inner_slice = make_partial_u8_slice(self.inner.as_ref(), 0, 8);
+
+        let result = format!("Block {{ inner: {inner_slice}, dirty: {} }}", self.dirty());
+
+        write!(f, "{result}")
+    }
+}
+
+impl Block {
+    /// Create a block with `capacity`, which usually equals the `block_size` of storage manager.
+    #[must_use]
+    pub fn new(capacity: usize) -> Self {
+        Block {
+            inner: Arc::new(AlignedBytes::new_zeroed(capacity, PAGE_SIZE)),
+            dirty: false,
+        }
+    }
+
+    /// Returns the length of the block, which usually equals the `block_size` of storage manager.
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Check if the block is with size of 0.
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Get a mutable slice of the underlying data, copy them if there are other blocks hold the same data with `Arc`.
+    /// See also [`Arc::make_mut`](fn@std::sync::Arc::make_mut).
+    pub fn make_mut(&mut self) -> &mut [u8] {
+        Arc::make_mut(&mut self.inner).as_mut()
+    }
+
+    /// Checks if the block is dirty.
+    #[must_use]
+    pub fn dirty(&self) -> bool {
```

Comment:
`is_empty` and `is_dirty`? or `empty` and `dirty`?

Reply:
The `dirty` method is a getter of a private field, and it's a convention to name a getter in the same name as the field it gets in Rust.

The `is_empty` method is not a getter, and in fact, it's implemented to satisfy the clippy lints `clippy::len_without_is_empty`.